README.md
# Clever Settings

[![style: very good analysis][very_good_analysis_badge]][very_good_analysis_link]
[![Powered by Mason](https://img.shields.io/endpoint?url=https%3A%2F%2Ftinyurl.com%2Fmason-badge)](https://github.com/felangel/mason)
[![License: MIT][license_badge]][license_link]
![Coverage](https://raw.githubusercontent.com/RoundedInfinity/clever_settings/fddb130e811980659044ae1b3bb86fed38f2f866/coverage_badge.svg)

Clever Settings is a Dart package that provides an easy way to manage and store settings for your application. It uses Hive, a lightweight and fast key-value store, to persist the settings data.

## Installation 💻

Add the following to your `pubspec.yaml`:

```yaml
dependencies:
  clever_settings: ^[version]
  # When using flutter (optional)
  clever_settings_flutter: ^[version]
```

Install it:

```sh
dart pub get
```

---

## Initialization 🚀

Before you can start using Clever Settings, you need to initialize [Hive](https://docs.hivedb.dev/#/README?id=initialize) first by calling `Hive.init` (or `Hive.initFlutter`). Call the open function once at the start of your application:

```dart
// Initialize Hive before this
await CleverSettings.open();
```

Alternatively, when using [clever_settings_flutter](https://pub.dev/packages/clever_settings_flutter), you can just call:

```dart
await CleverSettingsFlutter.init();
```

## Usage 💡

Clever Settings provides multiple classes for managing settings. SettingsValue is used for storing primitive types like bool, int, String, etc. SerializableSettingsValue is used for storing complex objects that can be serialized to and from JSON.

> For additional flutter features, see [clever_settings_flutter](https://pub.dev/packages/clever_settings_flutter).

### Example

```dart
class Settings {
  const Settings._();

  /// A settings value that uses a bool value
  static const darkMode =
      SettingsValue<bool>(name: 'darkMode', defaultValue: false);
}

...

void myMethod() {
  // This is automatically saved to disk.
  Settings.darkMode.value = true;

  // Get the value. 
  final darkModeEnabled = Settings.darkMode.value;
}
```

### SettingsValue

To create a new `SettingsValue`, simply provide a unique name for the setting and an optional default value:

```dart
final mySetting = SettingsValue<bool>(name: 'mySetting', defaultValue: true);
```

You can then get or set the value of the setting. This automatically stores the setting to storage:

```dart
final currentValue = mySetting.value;
mySetting.value = false;
```

You can also listen to changes to the setting by calling the `watch` function:

```dart
final stream = mySetting.watch();
stream.listen((newValue) {
  // do something with the new value
});
```

### SerializableSettingsValue

`SerializableSettingsValue` works the same way as `SettingsValue`, but you also need to provide `fromJson` and `toJson` functions to serialize and deserialize the value:

```dart
final myObjectSetting = SerializableSettingsValue<MyObject>(
  name: 'myObjectSetting',
  fromJson: (json) => MyObject.fromJson(json),
  toJson: (object) => object.toJson(),
);
```

### DefaultSettingsValue

DefaultSettingsValue is a subclass of SettingsValue that provides a default value if no value is specified by the user. The value stored in this class is guaranteed to be non-null.

```dart
final setting = SettingsValue<int>(name: 'setting', defaultValue: 34);
final defaultSetting = DefaultSettingsValue<int>(name: 'default', defaultValue: 16);

// This value is of type int? even though it has a default value.
final value = setting.value;

// This value is of type int and not nullable.
final otherValue = defaultSetting.value;
```

## License

This package is released under the MIT License. See LICENSE for more information.

[license_badge]: https://img.shields.io/badge/license-MIT-blue.svg
[license_link]: https://opensource.org/licenses/MIT
[very_good_analysis_badge]: https://img.shields.io/badge/style-very_good_analysis-B22C89.svg
[very_good_analysis_link]: https://pub.dev/packages/very_good_analysis


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.1.2+1 - 2023-05-06

### Added

- DefaultSettingsValue that returns non-nullable values
- `reset()` method to reset a SettingsValue to its `defaultValue`.

### Changed

- Improved readme and docs

## 0.1.1+1 - 2023-05-02

### Fixed

- fixed `SerializableSettingsValue` not working with `.watch`.

## 0.1.0+1 - 2023-04-30

### Added

- Initial version


test/src/clever_settings_test.dart
// ignore: lines_longer_than_80_chars
// ignore_for_file: prefer_const_constructors, cascade_invocations, inference_failure_on_function_invocation, unawaited_futures, avoid_equals_and_hash_code_on_mutable_classes
import 'dart:async';
import 'dart:io';

import 'package:clever_settings/clever_settings.dart';
import 'package:hive/hive.dart';
import 'package:test/test.dart';

void main() {
  setUp(() async {
    final path = Directory.current.path;
    Hive.init(path);
    await Hive.deleteFromDisk();
    await CleverSettings.open();
  });
  group('CleverSettings', () {
    test('value is null after load', () {
      final setting = SettingsValue<bool>(name: 'bool_setting');
      expect(setting.value, isNull);
    });

    test('defaultValue is null after load', () {
      final setting = SettingsValue<bool>(name: 'bool_setting');
      expect(setting.defaultValue, isNull);
    });

    test('settings a value to true', () {
      final setting = SettingsValue<bool>(name: 'bool_setting')..value = true;

      expect(setting.value, equals(true));
    });

    test('settings a value to false', () {
      final setting = SettingsValue<bool>(name: 'bool_setting')..value = false;

      expect(setting.value, equals(false));
    });

    test('using default value', () {
      final setting =
          SettingsValue<String>(name: 'string_setting', defaultValue: 'hey');

      expect(setting.value, equals('hey'));
    });

    test('resetting a default value', () {
      final setting =
          SettingsValue<String>(name: 'string_setting', defaultValue: 'hey');

      setting.value = 'ho';

      setting.reset();

      expect(setting.value, equals('hey'));
    });

    test('using same settings object value', () {
      final setting = SettingsValue<String>(name: 'some_value');
      final otherSetting = SettingsValue<String>(name: 'some_value');

      setting.value = 'hello';

      expect(setting.value, equals(otherSetting.value));
    });

    test('reset settings working', () async {
      final setting = SettingsValue<bool>(name: 'bool_setting');
      setting.value = true;
      await CleverSettings.resetSettings();

      expect(setting.value, isNull);
    });

    test('watch a setting', () async {
      final setting =
          SettingsValue<int>(name: 'streamed_setting', defaultValue: 0);

      final stream = setting.watch();

      expectLater(stream, emitsInOrder([1, 2, 3]));

      await Future<void>.delayed(Duration.zero);

      setting.value = 1;
      await Future<void>.delayed(Duration.zero);
      setting.value = 2;
      await Future<void>.delayed(Duration.zero);
      setting.value = 3;

      Hive.box(kSettingsBox).close();
    });

    test('watch a setting with null value', () async {
      final setting =
          SettingsValue<int>(name: 'streamed_setting', defaultValue: 0);

      final stream = setting.watch();

      expectLater(stream, emitsInOrder([null]));

      await Future<void>.delayed(Duration.zero);

      setting.value = null;

      unawaited(Hive.box(kSettingsBox).close());
    });

    test('serializable value', () {
      final setting = SerializableSettingsValue<User>(
        name: 'user',
        fromJson: User.fromJson,
        toJson: (value) => value.toJson(),
      );
      setting.value = User(name: 'John Pork', age: 27);

      expect(setting.value, isA<User>());
    });

    test('serializable value with default value', () {
      final user = User(name: 'peter', age: 99);
      final setting = SerializableSettingsValue<User>(
        name: 'user',
        fromJson: User.fromJson,
        toJson: (value) => value.toJson(),
        defaultValue: user,
      );

      expect(setting.value, equals(user));
    });

    test('serializable value null when decoding error', () {
      final setting = SerializableSettingsValue<User>(
        name: 'user',
        fromJson: BrokenUser.fromJson,
        toJson: (value) => value.toJson(),
      );

      setting.value = BrokenUser(name: 'peter', age: 99);

      expect(setting.value, isNull);
    });

    test('serializable value set to null', () {
      final setting = SerializableSettingsValue<User>(
        name: 'user',
        fromJson: User.fromJson,
        toJson: (value) => value.toJson(),
      );

      setting.value = null;

      expect(setting.value, isNull);
    });

    test('watch a serializable setting ', () async {
      final setting = SerializableSettingsValue<User>(
        name: 'watched_user',
        fromJson: User.fromJson,
        toJson: (value) => value.toJson(),
      );

      final users = [
        User(name: 'walter', age: 54),
        User(name: 'jesse', age: 25),
        User(name: 'GUS', age: 38),
      ];

      final stream = setting.watch();

      expectLater(stream, emitsInOrder(users));

      await Future<void>.delayed(Duration.zero);

      setting.value = users[0];
      await Future<void>.delayed(Duration.zero);
      setting.value = users[1];
      await Future<void>.delayed(Duration.zero);
      setting.value = users[2];
      Hive.box(kSettingsBox).close();
    });
  });

  test('get DefaultSettingsValue not nullable', () {
    final setting = DefaultSettingsValue<String>(
      name: 'defaulted_settings',
      defaultValue: 'hi',
    );

    expect(setting.value, isA<String>());
  });

  test('set DefaultSettingsValue not nullable', () {
    final setting = DefaultSettingsValue<String>(
      name: 'defaulted_settings',
      defaultValue: 'hi',
    );

    setting.value = null;

    expect(setting.value, equals('hi'));
  });
}

class BrokenUser extends User {
  BrokenUser({required super.name, required super.age});

  // ignore: avoid_unused_constructor_parameters
  factory BrokenUser.fromJson(Map<String, dynamic> map) {
    throw Error();
  }
}

class User {
  const User({
    required this.name,
    required this.age,
  });

  factory User.fromJson(Map<String, dynamic> map) {
    return User(
      name: map['name'] as String? ?? '',
      age: map['age'] as int? ?? 0,
    );
  }

  final String name;
  final int age;

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'age': age,
    };
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is User && other.name == name && other.age == age;
  }

  @override
  int get hashCode => name.hashCode ^ age.hashCode;

  @override
  String toString() => 'User(name: $name, age: $age)';
}


example/main.dart
// ignore_for_file: avoid_print, depend_on_referenced_packages

import 'dart:io';

import 'package:clever_settings/clever_settings.dart';
import 'package:hive/hive.dart';
import 'package:meta/meta.dart';

class Settings {
  const Settings._();

  /// A settings value that uses a bool value
  static const darkMode =
      SettingsValue<bool>(name: 'darkMode', defaultValue: false);

  /// A settings that contains a complex object
  static final user = SerializableSettingsValue<User>(
    name: 'user',
    fromJson: User.fromJson,
    toJson: (value) => value.toJson(),
    defaultValue: const User(name: 'John Pork', age: 27),
  );
}

void main(List<String> args) async {
  // init hive to store data. With flutter use Hive.initFlutter
  final path = Directory.current.path;
  Hive.init(path);

  await CleverSettings.open();

  print('Is dark mode on? ${Settings.darkMode.value}');

  // This is automatically saved to disk.
  Settings.darkMode.value = true;

  print('The user is ${Settings.user.value}');

  // This is also stored across restarts.
  Settings.user.value = const User(name: 'John Doe', age: 22);
}

/// A complex object that can be converted to json.
///
/// This should have equality when using the `.watch` method on a setting.
@immutable
class User {
  const User({
    required this.name,
    required this.age,
  });

  factory User.fromJson(Map<String, dynamic> map) {
    return User(
      name: map['name'] as String? ?? '',
      age: map['age'] as int? ?? 0,
    );
  }

  final String name;
  final int age;

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'age': age,
    };
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is User && other.name == name && other.age == age;
  }

  @override
  int get hashCode => name.hashCode ^ age.hashCode;
}


lib/src/clever_settings.dart
// ignore_for_file: comment_references, lines_longer_than_80_chars

import 'dart:async';
import 'dart:convert';

import 'package:clever_logger/clever_logger.dart';
import 'package:hive/hive.dart';

final _logger = CleverLogger(
  'Settings',
  shouldLog: CleverSettings.logging,
  logActions: [const ColorfulPrintAction(printer: EmojiPrinter())],
);

/// Constant value used for the settings box in Hive.
///
/// All settings are stored in this box.
const kSettingsBox = 'clever_settings';

/// Static class for functions to control settings
///
/// See also:
/// - [SettingsValue] to store and retrieve settings.
/// - [SerializableSettingsValue] to use complex object as your settings.
class CleverSettings {
  CleverSettings._(); // coverage:ignore-line

  /// Loads the storage in which the settings are saved.
  ///
  /// Should be called once before the app starts
  /// (typically in the main function).
  static Future<void> open() async {
    await Hive.openBox<dynamic>(kSettingsBox);
  }

  /// Enables logging for settings.
  static bool logging = true;

  /// Deletes all stored values for settings.
  static Future<void> resetSettings() async {
    await Hive.box<dynamic>(kSettingsBox).clear();
  }
}

/// {@template settings_value}
/// [SettingsValue] is a generic class that represents
/// a setting in an application.
///
/// It allows you to store and retrieve values
/// for a specific setting by its unique name.
///
/// You can store all primitive types: [List], [Map], [DateTime],
/// [BigInt] and [Uint8List].
/// To store other types, see [SerializableSettingsValue].
///
/// Example usage:
/// ```dart
/// final mySetting = SettingsValue<bool>(name:'my_setting',defaultValue: true);
/// // Set a new value for my_setting
/// mySetting.value = false;

/// // Retrieve the current value of my_setting
/// final currentValue = mySetting.value;

/// // Watch for changes to my_setting
/// final stream = mySetting.watch();
/// ```
/// {@endtemplate}
class SettingsValue<T> {
  /// Creates a new instance of a SettingsValue
  /// with the specified name and optional default value.
  ///
  /// {@macro settings_value}
  const SettingsValue({required this.name, this.defaultValue});

  /// The unique name of this setting.
  ///
  /// This is used as a key to store your values.
  /// Do not use the same name twice for different settings.
  ///
  /// Multiple [SettingsValue]s with the same name
  /// will return the identical value.
  final String name;

  /// The value that is used by default.
  ///
  /// When this is not set [SettingsValue] returns null.
  ///
  /// Unlike [value], [defaultValue] is not stored in storage
  /// but bound to this object.
  final T? defaultValue;

  /// The box used by this [SettingsValue]
  Box<dynamic> get _box => Hive.box(kSettingsBox);

  /// Get the value stored for [name].
  ///
  /// When no value is stored and [defaultValue] is set,
  /// this returns [defaultValue].
  T? get value {
    return _box.get(name, defaultValue: defaultValue) as T?;
  }

  /// Set the value stored for [name].
  set value(T? value) {
    _logger.logConfig('Setting $name changed to: $value');
    _box.put(name, value);
  }

  /// Resets the value of this [SettingsValue] to its default value.
  ///
  /// This sets the current value to the [defaultValue] specified
  /// when the instance was created.
  ///
  /// If no [defaultValue] was specified, the value of the [SettingsValue]
  /// will be set to null.
  void reset() {
    value = defaultValue;
  }

  /// Watches this [SettingsValue] for changes.
  Stream<T?> watch() {
    final stream = _box.watch(key: name);

    final transformer = StreamTransformer<BoxEvent, T?>.fromHandlers(
      handleData: (data, sink) {
        sink.add(data.value as T?);
      },
      handleDone: (sink) {
        sink.close();
      },
    );

    return stream.transform(transformer);
  }
}

/// {@template serializable_settings_value}
/// SerializableSettingsValue allows you to store and retrieve values
/// that can be serialized to and from JSON format.
///
/// See also:
/// - [SettingsValue]
/// {@endtemplate}
class SerializableSettingsValue<T> extends SettingsValue<T> {
  /// Creates a new instance of SerializableSettingsValue
  /// with the specified name, default value, and toJson and fromJson functions.
  ///
  /// {@macro serializable_settings_value}
  const SerializableSettingsValue({
    required super.name,
    required this.fromJson,
    required this.toJson,
    super.defaultValue,
  });

  /// Method that describes how to deserializes Json into an instance of [T].
  final T Function(Map<String, dynamic> json) fromJson;

  /// Method that describes how to serializes the value into Json.
  final Map<String, dynamic> Function(T value) toJson;

  T? _convertFromJson(dynamic data) {
    if (data == null) {
      return defaultValue;
    }

    assert(
      data is String,
      'Serializable Settings can only be stored as Strings',
    );

    try {
      return fromJson(jsonDecode(data as String) as Map<String, dynamic>);
    } catch (e) {
      _logger.logSevere(e);
      return null;
    }
  }

  @override
  T? get value {
    final data = _box.get(name);

    return _convertFromJson(data);
  }

  @override
  set value(T? value) {
    _logger.logConfig('Setting $name changed to: $value');

    if (value == null) {
      _box.put(name, null);
      return;
    }
    _box.put(name, jsonEncode(toJson(value)));
  }

  @override
  Stream<T?> watch() {
    final stream = _box.watch(key: name);

    final transformer = StreamTransformer<BoxEvent, T?>.fromHandlers(
      handleData: (data, sink) {
        sink.add(_convertFromJson(data.value));
      },
      handleDone: (sink) {
        sink.close();
      },
    );

    return stream.transform(transformer);
  }
}

/// {@template default_settings_value}
/// A settings value that provides a default value
/// if no value is specified by the user.
///
///The value stored in this class is guaranteed to be non-null.
///
/// Example:
/// ```
/// final setting = SettingsValue<int>(name: 'setting', defaultValue: 34);
/// final defaultSetting = DefaultSettingsValue<int>(name: 'default', defaultValue: 16);
///
/// // This value is of type int? even though it has a default value.
/// final value = setting.value;
///
/// // This value is of type int and not nullable.
/// final otherValue = defaultSetting.value;
///
/// ```
/// {@endtemplate}
class DefaultSettingsValue<T> extends SettingsValue<T> {
  /// {@macro default_settings_value}
  DefaultSettingsValue({required super.name, required T super.defaultValue});

  @override
  T get value => super.value!;

  /// Set the value stored for [name].
  ///
  /// When value is `null`, the settings' value is set to [defaultValue].
  @override
  set value(T? value) {
    super.value = value ?? defaultValue!;
  }
}
